### PR TITLE
#983 Wrapper Didcomm v2 in Swift

### PR DESCRIPTION
**Pull request by rodrigofranzoi**
Issue: https://github.com/sicpa-dlab/cbdc-projects/issues/983

### CODE REVIEW

File: ios/Extensions/DidDoc+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            fatalError("Cant resolve authentications")
+        }
+        
+        
+        guard let verificationMethodsJson = json["verification_methods"] as? [JSONDictionary] else {
+            fatalError("Cant resolve verificationMethodsJson")
+        }
+        
+        guard let servicesJson = json["services"] as? [JSONDictionary] else {
+            fatalError("Cant resolve verificationMethodsJson")
```
Comment (AlexanderShenshin) [inline ios/Extensions/DidDoc+Extensions.swift:22]: Error message should be `Can't resolve servicesJson`

Comment (AlexanderShenshin) [reply ios/Extensions/DidDoc+Extensions.swift:22]: Also, for `DidDoc` we have less detailed errors comparing to, for example, `VerificationMethod`:
`Cant resolve verificationMethodsJson`  vs `Cant resolve verification method: id`

It's better to use more or less same error message convention where possible.

File: ios/Source/Utils.swift [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+public func createResolvers(with resolversId: NSString) -> (DidResolverProxy, SecretsResolverProxy) {
+    let didResolver = DidResolverProxy(resolversProxyModule: DIDCommResolversProxyModule(), resolversId: resolversId.asString)
```
Comment (AlexanderShenshin) [inline ios/Source/Utils.swift:3]: Do we create new instances of `DIDCommResolversProxyModule` for every resolver proxy?

File: demo/src/App.tsx [hunk 3]
```diff
@@ -25,8 +25,11 @@ export default function App() {
 
   useEffect(() => {
     //Workaround for issue with non-working NativeEventEmitter if created from library code
-    const nativeEventEmitter = new NativeEventEmitter(DIDCommResolversProxyModule)
-    DIDCommResolversProxy.start(nativeEventEmitter)
+    const emitter =
```
Comment (AlexanderShenshin) [inline demo/src/App.tsx:28]: I still have the feeling that we can avoid using condition here. The problem with using `DIDCommResolversProxyModule` in iOS may be related to logic where every resolver proxy creates it's own instance of this module.

File: ios/Source/DIDCommResolversProxyModule.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    func setFoundSecret(_ jsonValue: NSString?, resolversId: String) {
+        NotificationCenter.default.post(name: NSNotification.Name(resolversId+"kid"), object: jsonValue)
+    }
+
+    @objc(setFoundSecretIds:resolversId:)
+    func setFoundSecretIds(_ jsonValue: NSString?, resolversId: String) {
+        NotificationCenter.default.post(name: NSNotification.Name(resolversId+"kids"), object: jsonValue)
+    }
+    
+    public func sendEvent(event: ResolverProxyEvent, resolversId: String) {
```
Comment (AlexanderShenshin) [inline ios/Source/DIDCommResolversProxyModule.swift:25]: Can be removed from instance and made static.

File: package.json [hunk 5]
```diff
@@ -31,39 +37,57 @@
...
-        }
-      ]
+      "module"
     ]
+  },
+  "eslintIgnore": [
+    "node_modules/",
+    "lib/"
+  ],
+  "dependencies": {
```
Comment (AlexanderShenshin) [inline package.json:86]: Double check if these are needed.

File: package.json [hunk 6]
```diff
@@ -2,10 +2,11 @@
   "name": "@sicpa-dlab/didcomm-react-native",
   "version": "0.0.9",
   "description": "React Native wrapper for DIDComm v2",
-  "main": "lib/module/index.js",
+  "main": "lib/commonjs/index.js",
```
Comment (AlexanderShenshin) [inline package.json:5]: Double check the reason of this change.

File: lefthook.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+pre-commit:
```
Comment (AlexanderShenshin) [inline lefthook.yml:1]: Not necessary.

File: ios/Source/Utils.swift (hunk 2 above)
Comment (AlexanderShenshin) [reply ios/Source/Utils.swift:3]: Decided to not store resolvers proxy module in `DidResolverProxy` and `SecretsResolverProxy` and, therefore, remove this instance creation.

File: ios/Source/DIDCommMessageHelpersModule.swift [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                                           forward: false,
+                                           forwardHeaders: [:],
+                                           messagingService: nil,
+                                           encAlgAuth: .a256cbcHs512Ecdh1puA256kw,
+                                           encAlgAnon: .xc20pEcdhEsA256kw)
+        
+        let (didResolver, secretsResolver) = createResolvers(with: resolversId)
+        let delegate = DidPromise(resolve, reject)
+        let _ = DidComm(didResolver: didResolver, secretResolver: secretsResolver)
+            .packEncrypted(msg: .init(fromJson: messageData),
```
Comment (AlexanderShenshin) [inline ios/Source/DIDCommMessageHelpersModule.swift:29]: Create a follow up ticket about error handling improvements.

File: ios/Extensions/Service+Extensions.swift [hunk 9]
```diff
@@ -0,0 +1,30 @@
+extension Service {
+    init(fromJson json: JSONDictionary) {
+        
+        print(json)
```
Comment (AlexanderShenshin) [inline ios/Extensions/Service+Extensions.swift:4]: Remove this.

File: demo/react-native.config.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+const path = require("path")
+
+module.exports = {
```
Comment (AlexanderShenshin) [inline demo/react-native.config.js:3]: Double check if this is needed.

File: ios/Source/DIDCommMessageHelpersModule.swift (hunk 8 above)
Comment (rodrigofranzoi) [reply ios/Source/DIDCommMessageHelpersModule.swift:29]: https://github.com/sicpa-dlab/cbdc-projects/issues/1051

File: demo/react-native.config.js (hunk 10 above)
Comment (rodrigofranzoi) [reply demo/react-native.config.js:3]: needed for both ios and android

File: android/src/main/java/com/sicpa/didcomm/reactnative/ResolversProxyModule.kt [hunk 11]
```diff
@@ -43,23 +43,23 @@ class ResolversProxyModule(private val reactContext: ReactApplicationContext) :
         get() = _foundSecretIdsChannel as ReceiveChannel<Set<String>>
 
     @ReactMethod
-    fun setResolvedDid(jsonValue: String?) {
+    fun setResolvedDid(jsonValue: String?, setResolvedDid: String) {
```
[the remark this answers is not in the document]
Comment (rodrigofranzoi) [reply android/src/main/java/com/sicpa/didcomm/reactnative/ResolversProxyModule.kt]: issue created here: https://github.com/sicpa-dlab/cbdc-projects/issues/1053

**Review by AlexanderShenshin — APPROVED** [review]

**Review by Artemkaaas — APPROVED** [review]

**Pull request merged**